pkg/pgmodel/ingestor/inserter.go

```diff
@@ -0,0 +1,378 @@
+		conn:             conn,
+		input:            input,
+		pending:          pendingBuffers.Get().(*pendingBuffer),
+		seriesCache:      make(map[string]utils.SeriesID),
+		seriesCacheEpoch: -1,
+		// set to run at half our deletion interval
+		seriesCacheRefresh: time.NewTicker(30 * time.Minute),
+		metricTableName:    tableName,
+		toCopiers:          toCopiers,
+	}
+
+	handler.handleReq(firstReq)
+
+	// Grab new requests from our channel and handle them. We do this hot-load
+	// style: we keep grabbing requests off the channel while we can do so
+	// without blocking, and flush them to the next layer when we run out, or
+	// reach a predetermined threshold. The theory is that wake/sleep and
+	// flushing is relatively expensive, and can be easily amortized over
+	// multiple requests, so it pays to batch as much as we are able. However,
+	// writes to a given metric can be relatively rare, so if we don't have
+	// additional requests immediately we're likely not going to for a while.
+	for {
+		if !handler.hasPendingReqs() {
+			stillAlive := handler.blockingHandleReq()
+			if !stillAlive {
+				return
+			}
+			continue
+		}
+
+	hotReceive:
+		for handler.nonblockingHandleReq() {
+			if len(handler.pending.batch.SampleInfos) >= flushSize {
+				break hotReceive
+			}
+		}
+
+		handler.flush()
+	}
+}
+
+// Create the metric table for the metric we handle, if it does not already
+// exist. This only does the most critical part of metric table creation, the
+// rest is handled by completeMetricTableCreation().
+func initializeInserterRoutine(
+	conn utils.PgxConn,
+	metricName string,
+	completeMetricCreationSignal chan struct{},
+	metricTableNames utils.MetricCache,
+) (tableName string, err error) {
+	tableName, err = metricTableNames.Get(metricName)
+	if err == utils.ErrEntryNotFound {
+		var possiblyNew bool
+		tableName, possiblyNew, err = utils.GetMetricTableName(conn, metricName)
+		if err != nil {
+			return "", err
+		}
+
+		//ignone error since this is just an optimization
+		_ = metricTableNames.Set(metricName, tableName)
+
+		if possiblyNew {
+			//pass a signal if there is space
+			select {
+			case completeMetricCreationSignal <- struct{}{}:
+			default:
+			}
+		}
+	} else if err != nil {
+		return "", err
+	}
+	return tableName, err
+}
+
+// Handles actual insertion into the DB.
+// We have one of these per connection reserved for insertion.
+func runInserter(conn utils.PgxConn, in chan copyRequest) {
+	// We grab copyRequests off the channel one at a time. This, and the name is
+	// a legacy from when we used CopyFrom to perform the insetions, and may
+	// change in the future.
+	for {
+		req, ok := <-in
+		if !ok {
+			return
+		}
+		err := doInsert(conn, req)
+		if err != nil {
+			err = insertErrorFallback(conn, req, err)
+		}
+
+		req.data.reportResults(err)
+		pendingBuffers.Put(req.data)
+	}
+}
+
+// certain errors are recoverable, handle those we can
+//   1. if the table is compressed, decompress and retry the insertion
```

Comment:
this comment line can be removed, since we mention the decompress/compress recovery method, but nothing about recovery specifics is present in the code. The `tryRecovery` function properly mentions that detail there